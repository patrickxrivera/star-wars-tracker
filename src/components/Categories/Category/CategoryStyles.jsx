import styled, { css } from 'styled-components';

import {
  getBorderBottomFrom,
  getOpacityFrom,
  getHoverOpacityFrom
} from './helpers/categoryStylesHelpers';

const ContainerStyles = css`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  color: ${props => props.color};
  margin: 0 3rem;
  padding-bottom: 2rem;
  border-bottom: ${props => getBorderBottomFrom({ ...props })};
  opacity: ${props => getOpacityFrom({ ...props })};

  &:hover {
    opacity: ${props => getHoverOpacityFrom({ ...props })};
    cursor: pointer;
  }
`;

const CategoryName = styled.span`
  font-size: 1.25rem;
  margin-top: 0.6rem;
`;

export { ContainerStyles, CategoryName };
