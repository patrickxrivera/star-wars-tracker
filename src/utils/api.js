import axios from 'axios';

import { endpoint, count } from '../components/Results/ResultsCardContainer/ResultsCardData';

const api = {
  async getDataModelFor(selected) {
    const results = await this._fetchResultsFor(selected);

    switch (selected) {
      case 'People':
        const peopleDataModel = this._createPeopleDataModelFrom(results);
        return peopleDataModel;
      default:
        throw new Error(`${selected} is an invalid value for the api endpoint. Please try again.`);
    }
  },

  async _fetchResultsFor(selected) {
    const targetEndpoint = endpoint[selected];
    const response = await axios.get(targetEndpoint);
    const results = response.data.results;
    return results;
  },

  async _createPeopleDataModelFrom(list) {
    let dataModel = {};
    const shortenedList = list.slice(0, count);

    const homeworldData = await this._getDataFrom(shortenedList, 'homeworld');
    const speciesData = await this._getDataFrom(shortenedList, 'species');

    dataModel = await this._addHomeworldDataToDataModel(homeworldData, dataModel);
    dataModel = await this._addSpeciesDataToDataModel(speciesData, dataModel);
    dataModel = this._addNamesToDataModel(shortenedList, dataModel);
    dataModel = this._formatPopulation(dataModel);

    return dataModel;
  },

  async _getDataFrom(list, target) {
    // TODO figure out how to make concurrent axios requests with axios.all()
    const endpoints = list.map((person) => axios.get(person[target]));
    const responses = await axios.all(endpoints);
    return responses;
  },

  async _addHomeworldDataToDataModel(homeworldData, dataModel) {
    homeworldData.forEach((response, idx) => {
      const homeworld = response.data.name;
      const population = response.data.population;

      dataModel[idx] = dataModel[idx] || {};
      dataModel[idx].Homeworld = homeworld;
      dataModel[idx].Population = population;
    });

    return dataModel;
  },

  async _addSpeciesDataToDataModel(speciesData, dataModel) {
    speciesData.forEach((response, idx) => {
      const species = response.data.name;
      const language = response.data.language;

      dataModel[idx] = dataModel[idx] || {};
      dataModel[idx].Species = species;
      dataModel[idx].Language = language;
    });

    return dataModel;
  },

  _addNamesToDataModel(list, dataModel) {
    list.forEach((person, idx) => {
      dataModel[idx] = dataModel[idx] || {};
      dataModel[idx].Name = person.name;
    });
    return dataModel;
  },

  _formatPopulation(dataModel) {
    for (let key in dataModel) {
      const formattedPopulation = Number(dataModel[key].Population).toLocaleString();
      dataModel[key].Population = formattedPopulation;
    }

    return dataModel;
  }
};

export default api;
