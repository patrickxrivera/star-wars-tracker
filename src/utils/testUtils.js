const sel = (id) => {
  return `[data-test="${id}"]`;
};

const simulateClick = (element) => {
  return element.simulate('click');
};

export { sel, simulateClick };
