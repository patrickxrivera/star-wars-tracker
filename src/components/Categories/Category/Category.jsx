import React, { Component } from 'react';
import styled from 'styled-components';
import { ContainerStyles, CategoryName } from './CategoryStyles.jsx';
import { func, string } from 'prop-types';

const Container = styled.div`
  ${ContainerStyles};
`;

const Category = ({ onClick, categoryName, color, Icon }) => (
  <Container
    onClick={() => onClick(categoryName)}
    key={categoryName}
    color={color}
  >
    <Icon size={35} />
    <CategoryName>{categoryName}</CategoryName>
  </Container>
);

Category.propTypes = {
  onClick: func.isRequired,
  categoryName: string.isRequired,
  color: string.isRequired,
  Icon: func.isRequired
};

export default Category;
