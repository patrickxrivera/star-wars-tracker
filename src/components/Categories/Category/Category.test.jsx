import React from 'react';
import { shallow, mount } from 'enzyme';
import sinon from 'sinon';
import { sel, simulateClick } from '../../../utils/testUtils';

import Category from './Category.jsx';

test('Category renders correctly', () => {
  const wrapper = shallow(<Category />);
  expect(wrapper).toMatchSnapshot();
});

// test('Calls handleClick when a category is clicked', () => {
//   const handleClick = sinon.spy();
//   const wrapper = shallow(<Category onClick={handleClick} />);
//   const targetEl = wrapper.find(sel('category-Vehicles')).first();
//   simulateClick(targetEl);
//
//   expect(handleClick.calledOnce).toEqual(true);
// });
