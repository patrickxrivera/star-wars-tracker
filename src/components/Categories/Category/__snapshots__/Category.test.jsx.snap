// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Category renders correctly 1`] = `
<styled.div
  data-test="category-undefined"
>
  <styled.span />
</styled.div>
`;
